### PR TITLE
Refactor how ScopeId->FunctionInfo is found in LiveTab

### PR DESCRIPTION
**Pull request by vwbaker**
Move scope_id_to_function_info_ into ScopeIdProvider; all of the required fields for function_info are already there and shouldn't need to be recalculated. GetAllProvidedScopeIds should be a superset of instrumented_functions in CaptureData so we aren't losing any.

This will also help future PRs as my goal here is to simplify where LiveTab gets its data so it can act on a smaller data set rather than all of CaptureData.

Bug: http://b/240112049

### CODE REVIEW

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h [hunk 1]
```diff
@@ -61,6 +60,13 @@ class LiveFunctionsDataView : public DataView {
 
   std::string GetToolTip(int /*row*/, int column) override;
 
+  void AddToIndices(ScopeId scope_id) {
```
Comment (akopich) [inline src/DataViews/include/DataViews/LiveFunctionsDataView.h:63]: As it is a public method now, maybe call is `AddScope`?

File: src/ClientData/ScopeIdProvider.cpp [hunk 2]
```diff
@@ -35,17 +36,24 @@ std::unique_ptr<NameEqualityScopeIdProvider> NameEqualityScopeIdProvider::Create
...
 
   for (const auto& instrumented_function : instrumented_functions) {
     const ScopeId scope_id{instrumented_function.function_id()};
     const ScopeInfo scope_info(instrumented_function.function_name(),
                                ScopeType::kDynamicallyInstrumentedFunction);
     scope_id_to_info.emplace(scope_id, scope_info);
     scope_info_to_id.emplace(scope_info, scope_id);
+    scope_id_to_function_info.emplace(
+        scope_id,
+        FunctionInfo{instrumented_function.file_path(), instrumented_function.file_build_id(),
```
Comment (akopich) [inline src/ClientData/ScopeIdProvider.cpp:49]: Maybe use `std::piecewise_construct`?  Might hurt readability though. Other reviewer's opinion?

File: src/DataViews/LiveFunctionsDataView.cpp [hunk 3]
```diff
@@ -478,56 +475,18 @@ void LiveFunctionsDataView::DoFilter() {
...
   UpdateHistogramWithScopeIds({});
-  scope_id_to_function_info_.clear();
   indices_.clear();
 
   if (!app_->HasCaptureData()) {
     DataView::OnDataChanged();
     return;
   }
 
-  const absl::flat_hash_map<uint64_t, orbit_grpc_protos::InstrumentedFunction>&
```
Comment (akopich) [inline src/DataViews/LiveFunctionsDataView.cpp:497]: Nice to see the chunk of code gone.

**Review by akopich — COMMENTED** [review]
Thanks for the nice change! 

I've left some remarks.

File: src/ClientData/include/ClientData/ScopeIdProvider.h [hunk 4]
```diff
@@ -80,6 +87,7 @@ class NameEqualityScopeIdProvider : public ScopeIdProvider {
   ScopeId max_instrumented_function_id_{};
   absl::flat_hash_map<const ScopeInfo, ScopeId> scope_info_to_id_ ABSL_GUARDED_BY(mutex_);
   absl::flat_hash_map<ScopeId, const ScopeInfo> scope_id_to_info_ ABSL_GUARDED_BY(mutex_);
+  absl::flat_hash_map<ScopeId, const FunctionInfo> scope_id_to_function_info_;
```
[the remark this answers is not in the document]
Comment (vwbaker) [reply src/ClientData/include/ClientData/ScopeIdProvider.h:91]: Done. I think as it's a larger in its own change I can do it separately, but will probably ask some clarifying questions about it.

File: src/DataViews/include/DataViews/LiveFunctionsDataView.h (hunk 1 above)
Comment (vwbaker) [reply src/DataViews/include/DataViews/LiveFunctionsDataView.h]: Done.

File: src/ClientData/ScopeIdProvider.cpp (hunk 2 above)
Comment (vwbaker) [reply src/ClientData/ScopeIdProvider.cpp:50]: https://en.cppreference.com/w/cpp/container/map/try_emplace afaict try_emplace already does this so I can't use try_emplace&piecewise_construct together?

File: src/ClientData/ScopeIdProvider.cpp [hunk 5]
```diff
@@ -35,17 +36,24 @@ std::unique_ptr<NameEqualityScopeIdProvider> NameEqualityScopeIdProvider::Create
...
   absl::flat_hash_map<const ScopeInfo, ScopeId> scope_info_to_id;
+  absl::flat_hash_map<ScopeId, const FunctionInfo> scope_id_to_function_info;
 
   for (const auto& instrumented_function : instrumented_functions) {
     const ScopeId scope_id{instrumented_function.function_id()};
     const ScopeInfo scope_info(instrumented_function.function_name(),
                                ScopeType::kDynamicallyInstrumentedFunction);
     scope_id_to_info.emplace(scope_id, scope_info);
     scope_info_to_id.emplace(scope_info, scope_id);
+    scope_id_to_function_info.emplace(
```
[the remark this answers is not in the document]
Comment (vwbaker) [reply src/ClientData/ScopeIdProvider.cpp]: Done.

File: src/ClientData/ScopeIdProvider.cpp (hunk 2 above)
Comment (akopich) [reply src/ClientData/ScopeIdProvider.cpp:50]: Oh, didn't know that. But this makes it even better. You can write 

```
scope_id_to_function_info.try_emplace(
        scope_id, instrumented_function.file_path(), instrumented_function.file_build_id(),
        instrumented_function.function_virtual_address(), instrumented_function.function_size(),
        instrumented_function.function_name());
```
This way we instantiate `FunctionInfo` only once. 
But then again, I am not sure if it's readable enough. I wonder what you and @antonrohr think.

**Review by akopich — APPROVED** [review]
LGTM! 

The only nit left to sort out is the `scope_id_to_function_info.try_emplace` call.

File: src/ClientData/ScopeIdProvider.cpp (hunk 2 above)
Comment (vwbaker) [reply src/ClientData/ScopeIdProvider.cpp]: Ah makes sense, thanks.  What do you think of a comment like this for readability?

File: src/ClientData/ScopeIdProvider.cpp [hunk 6]
```diff
@@ -46,10 +46,9 @@ std::unique_ptr<NameEqualityScopeIdProvider> NameEqualityScopeIdProvider::Create
     scope_id_to_info.try_emplace(scope_id, scope_info);
     scope_info_to_id.try_emplace(scope_info, scope_id);
     scope_id_to_function_info.try_emplace(
-        scope_id,
-        FunctionInfo{instrumented_function.file_path(), instrumented_function.file_build_id(),
-                     instrumented_function.function_virtual_address(),
-                     instrumented_function.function_size(), instrumented_function.function_name()});
+        scope_id, /* piecewise_construct FunctionInfo */ instrumented_function.file_path(),
```
Comment (akopich) [inline src/ClientData/ScopeIdProvider.cpp:49]: I'm afraid, `piecewise_construct` isn't the correct term here. `piecewise_construct` refers to e.g. constructing both key and the value. Here I'd rather say, we construct FunctionInfo in-place. So I suggest to replace the comment with "in-place FunctionInfo construction".

File: src/ClientData/ScopeIdProvider.cpp (hunk 2 above)
Comment (akopich) [reply src/ClientData/ScopeIdProvider.cpp]: https://github.com/google/orbit/pull/4208#discussion_r974145391

File: src/ClientData/ScopeIdProvider.cpp (hunk 6 above)
Comment (vwbaker) [reply src/ClientData/ScopeIdProvider.cpp]: Done

**Pull request merged**